### PR TITLE
check atomic and atomicapp version before nulecule app deploy

### PR DESCRIPTION
**Pull request by sub-mod**
Covers 2 issues
https://github.com/cockpit-project/cockpit/issues/2558
https://github.com/cockpit-project/cockpit/issues/2559

### CODE REVIEW

**Comment by stefwalter** [comment]
This page indicates that for validation errors, the dialog buttons should remain clickable, and any errors should be displayed after the user clicks the main ('Deploy') button.

https://github.com/cockpit-project/cockpit/wiki/HIG:-Dialogs

**Comment by sub-mod** [comment]
@stefwalter kindly review

File: pkg/kubernetes/nulecule.js [hunk 1]
```diff
@@ -85,6 +87,98 @@ define([
...
+            debug("checking nulecule specversion");
+            var process = cockpit.spawn(["/usr/bin/atomic", "info", image])
+                .fail(function(ex) {
+                    console.warn(ex);
+                    var message = _("Unable to fetch specversion from image.");
+                    var err = new Error(message);
+                    dfd.reject(err);
+                })
+                .done(function(data) {
+                    var datal = data.split("\n");
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: Could you split this code out into a helper function? It's very repetitive and makes this function very long.

File: pkg/kubernetes/nulecule.js [hunk 2]
```diff
@@ -85,6 +87,98 @@ define([
...
+                            msgl = datal[i].split('io.projectatomic.nulecule.providers:');
+                            if(msgl.length > 1)
+                                versions.providers = msgl[1];
+                        }
+                    }
+                    debug(versions);
+                    if (Object.keys(versions).length == 3) {
+                        dfd.resolve(versions);
+                    } else {
+                        var message;
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: Shouldn't we just say "This image is not a supported Nulecule image"? Perhaps we could log more information for people who want to debug the issue. But not show all this detail in the UI.

File: pkg/kubernetes/nulecule.js [hunk 3]
```diff
@@ -85,6 +87,98 @@ define([
...
 
+        function check_nulecule_version(image) {
+            var dfd = $.Deferred();
+            var msgl;
+            var versions = {};
+            debug("checking nulecule specversion");
+            var process = cockpit.spawn(["/usr/bin/atomic", "info", image])
+                .fail(function(ex) {
+                    console.warn(ex);
+                    var message = _("Unable to fetch specversion from image.");
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: This happens on my system if I type the following:

 * Type: Nulecule Application
 * Nulecule: projectatomic/helloapache
 * Namespace: projectatomic

Shouldn't we handle this case better with a more descriptive message, if people will run into it often?

File: pkg/kubernetes/nulecule.js [hunk 4]
```diff
@@ -85,6 +87,98 @@ define([
...
+
+        function check_atomic_version() {
+            var dfd = $.Deferred();
+            var msgl;
+            var versions = {};
+            debug("checking atomic version");
+            var process = cockpit.spawn(["/usr/bin/atomic", "--version"],{ err: "out", superuser: true , pty: true })
+                .fail(function(ex) {
+                    console.warn(ex);
+                    var message = _("atomic is not installed.");
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: This is way to abstract. Lets say: "The 'atomic' command is not installed on the system."

File: pkg/kubernetes/nulecule.js [hunk 5]
```diff
@@ -85,6 +87,98 @@ define([
...
+                    var err = new Error(message);
+                    dfd.reject(err);
+                })
+                .done(function(data) {
+                    var atomic_version = parseFloat(data);
+                    debug("atomic version is " + atomic_version);
+                    if (parseFloat(data) >= MINIMUM_SUPPORTED_ATOMIC_VERSION) {
+                        dfd.resolve(data);
+                    } else {
+                        var message = _("atomic version " + atomic_version + " is not supported.");
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: Translatable strings need to be constant. Please use cockpit.format() as you see it used elsewhere with translatable strings.

File: pkg/kubernetes/nulecule.js [hunk 6]
```diff
@@ -172,56 +266,78 @@ define([
...
-                                .done(function(content){
-                                    if(content) {
-                                        var jdata = JSON.parse(content);
-                                        debug(jdata);
-                                        self.answers = jdata;
-                                        deferred.resolve(jdata);
-                                    } else {
-                                        var message = _("Unable to read answer.conf.sample file.");
+            check_atomic_version()
+                .done(function() {
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: This entire block of code (the massive indentation) has become way way too unreadable and hard to follow. Please break it out into a few functions. If you need help, let me know and I can give more advice.

File: pkg/kubernetes/nulecule.js [hunk 7]
```diff
@@ -172,56 +266,78 @@ define([
...
-                                })
-                                .fail(function(ex){
-                                    console.warn(ex);
-                                    deferred.reject(ex);
-                                    file.close();
-                                });
+                                    });
+
+                            } else {
+                                var message = _("atomicapp version " + atomicapp_version + " is not supported.");
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: Again, this needs cockpit.format().

**Comment by stefwalter** [comment]
```
$ make
...
../pkg/kubernetes/nulecule.js:203:17: 'process' is not defined.
```

File: pkg/kubernetes/deploy.js [hunk 8]
```diff
@@ -380,9 +396,11 @@ define([
             $("#deploy-app-namespace").val('');
             $(".appentity").remove();
             nulecule_image.val('');
+            type_selector.prop('disabled',false);
```
Comment (stefwalter) [inline pkg/kubernetes/deploy.js]: This line is not needed.

File: pkg/kubernetes/deploy.js [hunk 9]
```diff
@@ -380,9 +396,11 @@ define([
...
             type_selector.val("manifest");
             type_selector.selectpicker('refresh');
 
+            $("#deploy-app-nulecule-image").prop("disabled", false);
```
Comment (stefwalter) [inline pkg/kubernetes/deploy.js]: This line is not needed.

**Comment by stefwalter** [comment]
We can remove those last two lines while merging.

**Comment by stefwalter** [comment]
Other than that looks good to me.

File: pkg/kubernetes/nulecule.js [hunk 10]
```diff
@@ -85,6 +87,121 @@ define([
...
+        }
+
+        function check_nulecule_version(image) {
+            var dfd = $.Deferred();
+            var versions = {};
+            debug("checking nulecule specversion");
+            var process = cockpit.spawn(["/usr/bin/atomic", "info", image])
+                .fail(function(ex) {
+                    console.warn(ex);
+                    var message = _("Unable to fetch LABELs from image.Please check if image is correct. ");
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: This needs rewording: "The image is not a correctly labeled Nulecule image."

File: pkg/kubernetes/nulecule.js [hunk 11]
```diff
@@ -85,6 +87,121 @@ define([
...
+                    var err = new Error(message);
+                    dfd.reject(err);
+                })
+                .done(function(data) {
+                    var atomic_version = parseFloat(data);
+                    debug("atomic version is " + atomic_version);
+                    if (parseFloat(data) >= MINIMUM_SUPPORTED_ATOMIC_VERSION) {
+                        dfd.resolve(data);
+                    } else {
+                        var message = cockpit.format(_("atomic version $0 is not supported."), atomic_version);
```
Comment (stefwalter) [inline pkg/kubernetes/nulecule.js]: Reworded: The 'atomic' command version $0 is not supported.
